STAC-23355 Fix every known and discovered issue post Dynatrace v2 API migration.

LouisParkin opened the pull request:
### Step 1: Link to Jira issue


### Step 2: Description of changes


### Step 3: Did you add / update tests for your changes in the right area?
- [ ] Unit/Component test
- [ ] Integration test	


### Step 4: I'm confident that everything is properly tested:
I got a PO / QA Approval by:
- [ ] Name

### Step 5: Can we ship this feature to production?
- [ ] Yes, I'm proud of my work. Ship it! :ship:

LouisParkin committed:
STAC-23355 Reduce overall 404 prints, fix broken component.

LouisParkin committed:
STAC-23355 Reduce overall 404 prints, fix broken component. [2]

dottorblaster commented on dynatrace_topology/stackstate_checks/dynatrace_topology/dynatrace_topology.py:
```diff
@@ -400,6 +404,15 @@ def _clean_unsupported_metadata(self, component):
                 component[key] = str(component[key])
                 self.log.debug('Converting %s from int to str.' % key)
+
+        # Handle nested properties
+        if "properties" in component and isinstance(component["properties"], dict):
+            properties = component["properties"]
+            # Handle releasesVersion field - convert string representation to empty dict if it's a string
+            if "releasesVersion" in properties and isinstance(properties["releasesVersion"], str):
+                self.log.debug('Converting releasesVersion from string representation to empty dict')
+                properties["releasesVersion"] = {}
+
```
I like the `if "x" in struct` statement, I might reuse that to improve the splunk integration

LouisParkin replied:
It improves api stability and data parseability.  Some of these fields are fluid and optional, making it hard to write a solid spec.

LouisParkin committed:
STAC-23355 Fix issue with osServices field.

LouisParkin committed:
STAC-23355 More robust handling of the osServices data element.

LouisParkin committed:
STAC-23355 More robust processing of the ProcessGroupEntity data element.

LouisParkin committed:
STAC-23355 Dealing with when customPgMetadata key is non-scalar.

LouisParkin committed:
STAC-23355 Dealing with when customPgMetadata key is non-scalar. [2]

LouisParkin committed:
STAC-23355 Don't create relations involving entity types that we don't create topology for.

LouisParkin committed:
STAC-23355 Fix tests broken by previous commit.

LouisParkin committed:
STAC-23355 Allow auth retry on error 401.

LouisParkin committed:
STAC-23355 Log file sanitization.

LouisParkin committed:
STAC-23355 Add a good-faith retry for http 5xx errors before failing.

LouisParkin committed:
Revert "STAC-23355 Add a good-faith retry for http 5xx errors before failing."

This reverts commit 3bfdcbd3e6d28a7a33f9a6c7a6e544ae4469043e.

LouisParkin committed:
STAC-23355 Dynatrace health performance updates.

LouisParkin committed:
STAC-23355 Add entity cache.

Potentially speed up processing if multiple events on the same entity.

LouisParkin committed:
STAC-23355 Use http session per check run when calling Dynatrace urls.

Instead of opening a new session every time - see if keep-alive improves performance.

LouisParkin committed:
STAC-23355 Added entity cache and warm-up + warm-up timing.

LouisParkin committed:
STAC-23355 Make cache warm-up optional, default enabled.  Print from-time for event data pull.

LouisParkin committed:
STAC-23355 Set from time to double the check interval if it was longer ago than that.

LouisParkin committed:
STAC-23355 Removed the "to"  parameter processing in the reply from dynatrace v2 api.

LouisParkin committed:
STAC-23355 Added checks-in-flight safeguard against additional checks starting up.

LouisParkin committed:
STAC-23355 Added unit tests for checks-in-flight safeguard.

LouisParkin committed:
STAC-23355 Silence log verbosity for better production readiness.

LouisParkin committed:
STAC-23355 Disable cache warm-up by default.

Now that the check doesn't fetch 5-days' worth of data anymore, not immediately beneficial to cache all entities in existence for 500-odd events.

LouisParkin committed:
STAC-23355 Unit tests for cache warming.

Fixed spacing for some of the log print statements that went from info -> debug.

craffit commented on dynatrace_base/stackstate_checks/dynatrace/dynatrace_client.py:
```diff
@@ -29,6 +30,18 @@ def __init__(self, token, verify=False, cert=None, keyfile=None, timeout=None, i
         self.timeout = timeout
         self.is_jwt_auth = is_jwt_auth
```
make this is_ms_jwt_auth

craffit commented on dynatrace_health/stackstate_checks/dynatrace_health/dynatrace_health.py:
```diff
@@ -21,13 +22,15 @@
     url: AnyUrlStr
     token: str
     instance_tags: List[str] = []
-    events_boostrap_days: int = EVENTS_BOOSTRAP_DAYS
```
fix constant name

craffit commented on dynatrace_health/stackstate_checks/dynatrace_health/dynatrace_health.py:
```diff
@@ -54,11 +60,64 @@ def get_health_stream(self, instance):
+
+            # Validate that timestamp isn't too old (more than double the check interval)
+            # This prevents processing too many events if state gets stale or corrupted
+            # Only applies if state already exists (not first run)
```
move comment down

LouisParkin committed:
STAC-23355 Address PR comments.
